T5 support

davidkaczer opened the pull request:
# What does this PR do?

This PR adds support for pretrained LongT5 encoder-decoder models from HuggingFace.

## General Changes
* Add HuggingFacePretrainedEncoderDecoderModel class for loading LongT5
* Add span masking collator ported from [official Google implementation](https://github.com/huggingface/transformers/blob/main/examples/flax/language-modeling/run_t5_mlm_flax.py)
* Add test for span masking collator
* Add example training configs for LongT5
* Add logic to distinguish between decoder-only and encoder-decoder models when passing inputs

## Breaking Changes
* None intended

## Checklist before submitting final PR
- [x] My PR is minimal and addresses one issue in isolation
- [x] I have merged the latest version of the target branch into this feature branch
- [x] I have reviewed my own code w.r.t. correct implementation, missing type hints, proper documentation, etc.
- [x] I have run a sample config for model training
- [ ] I have checked that all tests run through (`python tests/tests.py`) (apparently some tests were failing already in upstream? also didn't test multi-GPU)
- [ ] I have updated the internal changelog (`CHANGELOG_DEV.md`)

davidkaczer committed:
feat(t5): WIP: add support for T5 model

- add SpanMaskingCollateFn for span denoising objective
- support loading T5 checkpoint from huggingface
- support passing decoder inputs to T5 model
- add example config for pretraining T5 checkpoint from HF

davidkaczer committed:
fix(t5): shift target tokens for input to decoder

davidkaczer committed:
Merge branch 'Modalities:main' into t5

davidkaczer committed:
chore: add span masking collator test

davidkaczer committed:
Merge branch 't5' of https://github.com/davidkaczer/modalities into t5

Commit:
update training yaml file using T5 model and tokenizer, also create new huggingface model wrapper for encoder decoder architecture

davidkaczer committed:
fix(t5): encoder and decoder model handling logic

- add logic to model_predict_batch
- fix example config
- cleanup HF model implementation

davidkaczer committed:
chore: pre-commit hook compliance

davidkaczer committed:
docs: add docstrings and type hints

le1nux commented on config_files/training/config_lorem_ipsum_span_masking_T5.yaml:
```diff
@@ -0,0 +1,272 @@
+    target_key: ${settings.referencing_keys.target_key}
+    noise_density: 0.3
+    mean_noise_span_length: 3
+    tokenizer:
```
need to check if we really need the tokenizer dependency

davidkaczer replied:
Removed, now only pass vocab_size (int)

le1nux commented on config_files/training/config_lorem_ipsum_span_masking_T5.yaml:
```diff
@@ -0,0 +1,272 @@
+    eps: 1e-8
+    weight_decay: 1e-1
+    # need to fix in huggingface_model.py to same spec as in gpt2_model.py
+    weight_decay_groups_excluded: []
```
Need to check if this has to be filled

davidkaczer replied:
Fixed, logic in the model is already there, just changed the config

le1nux commented on examples/getting_started/tokenizer_t5/config.json:
```diff
@@ -0,0 +1,33 @@
+{
```
Maybe move this outside of getting_started?

davidkaczer replied:
Moved to data/

le1nux commented on src/modalities/models/huggingface/collator.py:
```diff
@@ -0,0 +1,226 @@
+        start_indices[:, 0] = mask_indices[:, 0]
+
+        sentinel_ids = np.where(start_indices != 0, np.cumsum(start_indices, axis=-1), start_indices)
+        sentinel_ids = np.where(sentinel_ids != 0, (self.tokenizer.vocab_size - sentinel_ids), 0)
```
we might want to replace the `tokenizer` parameter by the two parameters `eos_token_id` and `vocab_size`

davidkaczer replied:
Done. eos_token_id is no longer needed

le1nux commented on src/modalities/models/huggingface/collator.py:
```diff
@@ -0,0 +1,226 @@
+        # masked tokens coming after sentinel tokens and should be removed
+        input_ids = input_ids_full[input_ids_full >= 0].reshape((batch_size, -1))
+        input_ids = np.concatenate(
+            [input_ids, np.full((batch_size, 1), self.tokenizer.tokenizer.eos_token_id)], axis=-1, dtype=np.int32
```
Need to verify that we actually need the `eos_token_id` here. I would assume that the pbin file already comes with the eos_token for every document.

davidkaczer replied:
Turns out this adds an EOS token at the end of every batch, not document, which seems clearly bad. So removed it

le1nux commented on src/modalities/models/model.py:
```diff
@@ -72,6 +72,12 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
+    # Ideally we would do this by checking isinstance(model, HuggingFacePretrainedEncoderDecoderModel),
+    # but HuggingFacePretrainedEncoderDecoderModel depends on NNModel from this file
+    # so this creates a circular dependency
+    if hasattr(model, "huggingface_model") and hasattr(model.huggingface_model, "encoder"):
+        forward_result = model.forward(inputs=batch.samples, targets=batch.targets)
+    else:
+        forward_result = model.forward(inputs=batch.samples)
```
Can we get rid of this check and have the huggingface model also implement the NNModelIF? We could have a wrapper for the HuggingFace model in this case.

davidkaczer replied:
Do you mean NNModel? There is no class NNModelIF.

I am not convinced about adding yet another wrapper class. Can you elaborate on how you would approach this?

IMO we should either
- refactor so that model_predict_batch is a method of NNModel that its subclass HuggingFacePretrainedEncoderDecoderModel inherits and redefines (preferred)
- move model_predict_batch to models.utils and then check type using an import (less preferred because it doesn't create a circular import now but creates the risk of one in the future) 
- leave as is (ugly but effective)

davidkaczer replied:
I briefly tried 
- refactor so that model_predict_batch is a method of NNModel that its subclass HuggingFacePretrainedEncoderDecoderModel inherits and redefines (preferred)

and there is some strange behavior. The current implementation as a separate function somehow automatically transfers the batch tensors from CPU to GPU, and changing to NNModel method breaks this. Probably the FSDP wrapper is responsible.

le1nux replied:
I just remebered, that we already have an implementation for it. 
https://github.com/Modalities/modalities/blob/main/src/modalities/models/huggingface/huggingface_model.py#L43

I think it would work out-of-the-box with a minor refactoring maybe. 

There are two things which I think we should consider: 
1. In the current branch version, we have low-level if else cases in a high-level routine (`model_predict_batch`). That's quite ugly.
2. `model_predict_batch` was introduced such that the models themselves don't need to know what a batch is. They only have their weights and a forward routine based on an input dict (`Dict[str, torch.Tensor]`). Huggingface tends to place lots of logic into the models, such as targets, loss calculation etc. From our perspective a model must not have any knowledge regarding these things. 

Are we currently also calculating the loss outside of T5? If so, it should possible to easily integrate `HuggingFacePretrainedModel`.

davidkaczer replied:
Changed the type check to this:
```if model.model_type.value in [LongT5Model, LongT5ForConditionalGeneration]:```
using the `model_type` attribute  from https://github.com/Modalities/modalities/blob/main/src/modalities/models/huggingface/huggingface_model.py#L43

Is this what you had in mind?

davidkaczer committed:
refactor: t5 config files

davidkaczer committed:
fix: don't add EOS token to span masking batch

davidkaczer committed:
fix: set excluded weight groups in t5 config

davidkaczer committed:
fix: remove tokenizer dependency

- remove dependency in span masking collator
- refactor t5 config
- refactor collator test

davidkaczer committed:
refactor: cleaner typecheck in model_predict_batch

davidkaczer committed:
refactor: change parent class of HuggingFacePretrainedEncoderDecoderModel

- simplify future type checks

davidkaczer commented:
Also changed `HuggingFacePretrainedEncoderDecoderModel` to be a child of `HuggingFacePretrainedModel` instead of its sibling. This simplifies potential future type checks and avoids duplicating methods needlessly.
